File: Simple-E-Commerce/CheckoutService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Simple_E_Commerce
{
    internal interface CheckoutService
    {
        public static void Checkout(Customer customer)
        {
            var cart = customer.Cart;

            if (cart.IsEmpty())
                throw new Exception("Cart is Empty");

            double subtotal = 0;
            double shippingFees = 0;

            List<IShippable> shippableItems = new List<IShippable>();

            foreach(var e in cart.GetItem())
            {
                Product product = e.Key;
                int quantity = e.Value;

                if (quantity > product.Quantity)
                    throw new Exception($"Product : {product.Name} is out of stock.");

                if (product is IExpirable expire && expire.IsExpired())
                    throw new Exception($"Product : {product.Name} is Expired.");

                subtotal += (product.Price * product.Quantity);


                if (product is IShippable ishippable)
                {
                    shippableItems.Add(ishippable);
                    shippingFees += ishippable.GetWeight() * 1.5 ; 
                }

            }

            double total = subtotal + shippingFees;

            if (total > customer.Balance)
                throw new Exception("Insufficient balance.");

            foreach (var e in cart.GetItem())
                e.Key.ReduceQuantity(e.Value);

            customer.ReduceBalance(total);

            Console.WriteLine("\n=== Checkout Summary ===");
            Console.WriteLine($"Subtotal: ${subtotal}");
            Console.WriteLine($"Shipping Fees: ${shippingFees}");
            Console.WriteLine($"Total Paid: ${total}");
            Console.WriteLine($"Remaining Balance: ${customer.Balance}");

            if (shippableItems.Count > 0)
            {
                ShippingService shippingService = new();
                shippingService.ShippingItems(shippableItems);
            }

            cart.Clear(); 
        }
    }
}

File: Using_Java/Simple-E-Commerce/src/CheckoutService.java
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class CheckoutService {
    public static void checkout(Customer customer) throws Exception {
        Cart cart = customer.getCart();

        if (cart.isEmpty()) {
            throw new Exception("Cart is Empty");
        }

        double subtotal = 0;
        double shippingFees = 0;

        List<IShippable> shippableItems = new ArrayList<>();

        for (Map.Entry<Product, Integer> entry : cart.getItems().entrySet()) {
            Product product = entry.getKey();
            int quantity = entry.getValue();

            if (quantity > product.getQuantity()) {
                throw new Exception("Product : " + product.getName() + " is out of stock.");
            }

            if (product instanceof IExpirable) {
                IExpirable expire = (IExpirable) product;
                if (expire.isExpired()) {
                    throw new Exception("Product : " + product.getName() + " is Expired.");
                }
            }

            subtotal += (product.getPrice() * product.getQuantity());

            if (product instanceof IShippable) {
                IShippable shippable = (IShippable) product;
                shippableItems.add(shippable);
                shippingFees += shippable.getWeight() * 1.5;
            }
        }

        double total = subtotal + shippingFees;

        if (total > customer.getBalance()) {
            throw new Exception("Insufficient balance.");
        }

        for (Map.Entry<Product, Integer> entry : cart.getItems().entrySet()) {
            entry.getKey().reduceQuantity(entry.getValue());
        }

        customer.reduceBalance(total);

        System.out.println("\n=== Checkout Summary ===");
        System.out.println("Subtotal: $" + subtotal);
        System.out.println("Shipping Fees: $" + shippingFees);
        System.out.println("Total Paid: $" + total);
        System.out.println("Remaining Balance: $" + customer.getBalance());

        if (shippableItems.size() > 0) {
            ShippingService shippingService = new ShippingService();
            shippingService.shippingItems(shippableItems);
        }

        cart.clear();
    }
} 